File: lib/swimMonthlyChallenges.js
/**
 * Personalized monthly challenges derived from recent swim history.
 * Three challenges are generated at the start of each month (deterministic).
 * Complete 1 → bronze, 2 → silver, 3 → gold for that month.
 */

const CHALLENGE_TYPES = ['sessions', 'distance', 'kcal', 'streak', 'active_weeks'];

const roundUp = (value, step) => Math.ceil(value / step) * step;

const monthKeyFromDate = (dateStr) => dateStr.slice(0, 7);

export const getMonthKey = (date = new Date()) => {
  const y = date.getFullYear();
  const m = String(date.getMonth() + 1).padStart(2, '0');
  return `${y}-${m}`;
};

const hashMonth = (monthKey) => {
  let h = 0;
  for (let i = 0; i < monthKey.length; i += 1) {
    h = (h * 31 + monthKey.charCodeAt(i)) | 0;
  }
  return Math.abs(h);
};

const getWeekKey = (dateStr) => {
  const d = new Date(dateStr);
  const day = d.getDay();
  const diff = d.getDate() - day + (day === 0 ? -6 : 1);
  const monday = new Date(d);
  monday.setDate(diff);
  return monday.toISOString().slice(0, 10);
};

const maxConsecutiveDaysInMonth = (sessions, monthKey) => {
  const dates = [...new Set(
    sessions.filter((s) => s.date.startsWith(monthKey)).map((s) => s.date)
  )].sort();
  if (!dates.length) return 0;
  let max = 1;
  let streak = 1;
  for (let i = 1; i < dates.length; i += 1) {
    const prev = new Date(dates[i - 1]);
    const curr = new Date(dates[i]);
    const diff = Math.round((curr - prev) / 86400000);
    streak = diff === 1 ? streak + 1 : 1;
    max = Math.max(max, streak);
  }
  return max;
};

const activeWeeksInMonth = (sessions, monthKey) => {
  const weeks = new Set(
    sessions.filter((s) => s.date.startsWith(monthKey)).map((s) => getWeekKey(s.date))
  );
  return weeks.size;
};

export const computeRecentMonthlyStats = (sessions, beforeMonthKey) => {
  const byMonth = {};
  sessions.forEach((s) => {
    const mk = monthKeyFromDate(s.date);
    if (mk >= beforeMonthKey) return;
    if (!byMonth[mk]) byMonth[mk] = { sessions: 0, distanceM: 0, activeKcal: 0, streak: 0 };
    byMonth[mk].sessions += 1;
    byMonth[mk].distanceM += s.metrics?.distanceM || 0;
    byMonth[mk].activeKcal += s.metrics?.activeKcal || 0;
  });

  const months = Object.keys(byMonth).sort().slice(-3);
  if (!months.length) {
    return { avgSessions: 4, avgDistance: 6000, avgKcal: 2500, avgStreak: 2, avgWeeks: 2 };
  }

  const totals = months.reduce(
    (acc, mk) => {
      acc.sessions += byMonth[mk].sessions;
      acc.distanceM += byMonth[mk].distanceM;
      acc.activeKcal += byMonth[mk].activeKcal;
      return acc;
    },
    { sessions: 0, distanceM: 0, activeKcal: 0 }
  );

  const n = months.length;
  const streaks = months.map((mk) => maxConsecutiveDaysInMonth(sessions, mk));
  const weeks = months.map((mk) => activeWeeksInMonth(sessions, mk));

  return {
    avgSessions: Math.max(1, Math.round(totals.sessions / n)),
    avgDistance: Math.max(1000, Math.round(totals.distanceM / n)),
    avgKcal: Math.max(500, Math.round(totals.activeKcal / n)),
    avgStreak: Math.max(1, Math.round(streaks.reduce((a, b) => a + b, 0) / n)),
    avgWeeks: Math.max(1, Math.round(weeks.reduce((a, b) => a + b, 0) / n)),
  };
};

/**
 * `intensity` scales challenge difficulty per coach mascot:
 * Flip 0.75 (easy), Flo 1 (balanced), Fins 1.25 (hard).
 */
const buildTarget = (type, stats, tierIndex, intensity = 1) => {
  const stretch = (1 + tierIndex * 0.12) * intensity;
  const intensityStep = intensity > 1 ? 1 : intensity < 1 ? -1 : 0;
  switch (type) {
    case 'sessions':
      return Math.max(2, roundUp(stats.avgSessions * stretch, 1));
    case 'distance':
      return Math.max(roundUp(4000 * intensity, 500), roundUp(stats.avgDistance * stretch, 500));
    case 'kcal':
      return Math.max(roundUp(1500 * intensity, 250), roundUp(stats.avgKcal * stretch, 250));
    case 'streak':
      return Math.max(2, Math.min(7, stats.avgStreak + tierIndex + intensityStep));
    case 'active_weeks':
      return Math.max(2, Math.min(4, stats.avgWeeks + tierIndex - 1 + intensityStep));
    default:
      return 1;
  }
};

const pickChallengeTypes = (monthKey) => {
  const seed = hashMonth(monthKey);
  const pool = [...CHALLENGE_TYPES];
  const picked = [];
  let s = seed || 1;
  for (let i = 0; i < 3; i += 1) {
    s = (s * 1103515245 + 12345) & 0x7fffffff;
    const at = s % pool.length;
    picked.push(pool.splice(at, 1)[0]);
  }
  return picked;
};

const pickRerollType = (monthKey, tierIndex, candidates, salt = 0) => {
  if (!candidates.length) return null;
  let s = hashMonth(`${monthKey}:reroll:${tierIndex}:${salt}`) || 1;
  s = (s * 1103515245 + 12345) & 0x7fffffff;
  return candidates[s % candidates.length];
};

export const normalizeMonthRerollEntry = (entry) => {
  if (!entry || typeof entry !== 'object') {
    return { overrides: {}, freeUses: 0 };
  }
  if (typeof entry.tierIndex === 'number' && entry.type) {
    return { overrides: { [entry.tierIndex]: entry.type }, freeUses: 1 };
  }
  const freeUses = typeof entry.freeUses === 'number'
    ? Math.max(0, entry.freeUses)
    : entry.freeUsed ? 1 : 0; // legacy boolean field
  return {
    overrides: entry.overrides && typeof entry.overrides === 'object' ? { ...entry.overrides } : {},
    freeUses,
  };
};

export const normalizeMonthlyChallengeRerolls = (raw) => {
  if (!raw || typeof raw !== 'object') return {};
  return Object.fromEntries(
    Object.entries(raw).map(([monthKey, entry]) => [monthKey, normalizeMonthRerollEntry(entry)])
  );
};

export const getMonthRerollOverrides = (monthKey, rerolls = {}) => (
  normalizeMonthRerollEntry(rerolls[monthKey]).overrides
);

export const getFreeMonthlyRerollsUsed = (monthKey, rerolls = {}) => (
  normalizeMonthRerollEntry(rerolls[monthKey]).freeUses
);

export const hasRerollAvailability = (monthKey, rerolls = {}, credits = 0, freeLimit = 1) => (
  getFreeMonthlyRerollsUsed(monthKey, rerolls) < freeLimit || credits > 0
);

const applyRerollOverrides = (challenges, monthEntry, stats, intensity = 1) => {
  const { overrides } = normalizeMonthRerollEntry(monthEntry);
  return challenges.map((ch, i) => {
    const type = overrides[i];
    if (!type) return ch;
    return {
      id: `${ch.monthKey}_${type}`,
      type,
      monthKey: ch.monthKey,
      target: buildTarget(type, stats, i, intensity),
      tierIndex: i,
    };
  });
};

export const generateMonthlyChallenges = (sessions, monthKey, rerolls = {}, intensity = 1) => {
  const stats = computeRecentMonthlyStats(sessions, monthKey);
  const types = pickChallengeTypes(monthKey);

  const challenges = types.map((type, i) => ({
    id: `${monthKey}_${type}`,
    type,
    monthKey,
    target: buildTarget(type, stats, i, intensity),
    tierIndex: i,
  }));

  return applyRerollOverrides(challenges, rerolls[monthKey], stats, intensity);
};

/** @returns {{ tierIndex: number, type: string } | null} */
export const createMonthlyChallengeReroll = (sessions, monthKey, tierIndex, rerolls = {}) => {
  if (tierIndex < 0 || tierIndex > 2) return null;
  const base = generateMonthlyChallenges(sessions, monthKey, rerolls);
  const lockedTypes = base.filter((_, i) => i !== tierIndex).map((ch) => ch.type);
  const currentType = base[tierIndex]?.type;
  const candidates = CHALLENGE_TYPES.filter(
    (type) => !lockedTypes.includes(type) && type !== currentType
  );
  const salt = Object.keys(getMonthRerollOverrides(monthKey, rerolls)).length;
  const type = pickRerollType(monthKey, tierIndex, candidates, salt);
  if (!type) return null;
  return { tierIndex, type };
};

export const hasMonthlyChallengeReroll = (monthKey, rerolls = {}) => (
  Object.keys(getMonthRerollOverrides(monthKey, rerolls)).length > 0
);

export const canRerollMonthlyChallenge = (
  sessions,
  monthKey,
  tierIndex,
  rerolls = {},
  credits = 0,
  { intensity = 1, freeLimit = 1 } = {}
) => {
  if (!hasRerollAvailability(monthKey, rerolls, credits, freeLimit)) return false;
  const state = evaluateMonthlyChallenges(sessions, monthKey, rerolls, intensity);
  const challenge = state.challenges[tierIndex];
  if (!challenge || challenge.completed) return false;
  return Boolean(createMonthlyChallengeReroll(sessions, monthKey, tierIndex, rerolls));
};

const measureChallenge = (type, monthSessions, monthKey) => {
  switch (type) {
    case 'sessions':
      return monthSessions.length;
    case 'distance':
      return monthSessions.reduce((s, x) => s + (x.metrics?.distanceM || 0), 0);
    case 'kcal':
      return monthSessions.reduce((s, x) => s + (x.metrics?.activeKcal || 0), 0);
    case 'streak':
      return maxConsecutiveDaysInMonth(monthSessions, monthKey);
    case 'active_weeks':
      return activeWeeksInMonth(monthSessions, monthKey);
    default:
      return 0;
  }
};

export const evaluateMonthlyChallenges = (sessions, monthKey = getMonthKey(), rerolls = {}, intensity = 1) => {
  const definitions = generateMonthlyChallenges(sessions, monthKey, rerolls, intensity);
  const monthSessions = sessions.filter((s) => s.date.startsWith(monthKey));

  const challenges = definitions.map((def) => {
    const current = measureChallenge(def.type, monthSessions, monthKey);
    const completed = current >= def.target;
    return { ...def, current, completed };
  });

  const completedCount = challenges.filter((c) => c.completed).length;
  let tier = null;
  if (completedCount >= 3) tier = 'gold';
  else if (completedCount >= 2) tier = 'silver';
  else if (completedCount >= 1) tier = 'bronze';

  const earnedAt = tier && monthSessions.length
    ? monthSessions[monthSessions.length - 1].date
    : null;

  return { monthKey, challenges, completedCount, tier, earnedAt };
};

export const tierRank = (tier) => ({ bronze: 1, silver: 2, gold: 3 }[tier] || 0);

export const getMonthlyTierUpgrade = (
  sessionsBefore,
  sessionsAfter,
  monthKey = getMonthKey(),
  rerolls = {},
  intensity = 1
) => {
  const before = evaluateMonthlyChallenges(sessionsBefore, monthKey, rerolls, intensity);
  const after = evaluateMonthlyChallenges(sessionsAfter, monthKey, rerolls, intensity);
  if (tierRank(after.tier) > tierRank(before.tier)) {
    return {
      monthKey,
      tier: after.tier,
      fromTier: before.tier,
      completedCount: after.completedCount,
      earnedAt: after.earnedAt,
    };
  }
  return null;
};

export const getMonthlyChallengeHistory = (
  sessions,
  { previewMonthlyMedals = false, monthlyChallengeRerolls = {}, intensity = 1 } = {}
) => {
  const months = [...new Set(sessions.map((s) => monthKeyFromDate(s.date)))].sort().reverse();
  const real = months
    .map((monthKey) => evaluateMonthlyChallenges(sessions, monthKey, monthlyChallengeRerolls, intensity))
    .filter((m) => m.tier);

  if (!previewMonthlyMedals) return real;

  const preview = getPreviewMonthlyMedalHistory(sessions);
  const realKeys = new Set(real.map((m) => m.monthKey));
  return [
    ...preview.filter((p) => !realKeys.has(p.monthKey)),
    ...real,
  ].sort((a, b) => b.monthKey.localeCompare(a.monthKey));
};

/** @internal Sample bronze / silver / gold months for UI testing. */
export const buildPreviewMonthlyMedal = (sessions, monthKey, tier) => {
  const state = evaluateMonthlyChallenges(sessions, monthKey);
  const completedCount = tier === 'gold' ? 3 : tier === 'silver' ? 2 : 1;
  const challenges = state.challenges.map((ch, i) => ({
    ...ch,
    completed: i < completedCount,
    current: i < completedCount ? ch.target : Math.max(0, Math.floor(ch.target * 0.4)),
  }));
  const earnedDay = String(Math.min(28, 10 + completedCount * 5)).padStart(2, '0');

  return {
    monthKey,
    challenges,
    completedCount,
    tier,
    earnedAt: `${monthKey}-${earnedDay}`,
    isPreview: true,
  };
};

export const getPreviewMonthlyMedalHistory = (sessions, date = new Date()) => {
  const tiers = ['gold', 'silver', 'bronze'];
  return tiers.map((tier, i) => {
    const d = new Date(date.getFullYear(), date.getMonth() - i, 1);
    return buildPreviewMonthlyMedal(sessions, getMonthKey(d), tier);
  });
};

/** One medal per calendar month for a year (tier null if not yet earned). */
export const getMonthlyMedalsForYear = (
  sessions,
  year = new Date().getFullYear(),
  monthlyChallengeRerolls = {},
  intensity = 1
) => {
  const months = [];
  for (let m = 1; m <= 12; m += 1) {
    const monthKey = `${year}-${String(m).padStart(2, '0')}`;
    const state = evaluateMonthlyChallenges(sessions, monthKey, monthlyChallengeRerolls, intensity);
    months.push({
      monthKey,
      month: m,
      tier: state.tier,
      completedCount: state.completedCount,
      challenges: state.challenges,
      earnedAt: state.earnedAt,
      hasSessions: sessions.some((s) => s.date.startsWith(monthKey)),
    });
  }
  return months;
};

/** Years that have at least one swim session, sorted descending. */
export const getSessionYears = (sessions) => {
  const years = new Set(
    sessions.map((s) => parseInt(s.date.slice(0, 4), 10)).filter(Boolean)
  );
  return [...years].sort((a, b) => b - a);
};

/** @internal */
export const __testing = {
  pickChallengeTypes,
  pickRerollType,
  computeRecentMonthlyStats,
  buildTarget,
  tierRank,
};

File: ios/AapSC/Lib/SwimMonthlyChallenges.swift
import Foundation

enum SwimMonthlyChallenges {
    private static let challengeTypes = ["sessions", "distance", "kcal", "streak", "active_weeks"]

    static func getMonthKey(_ date: Date = Date()) -> String {
        let calendar = Calendar.current
        let year = calendar.component(.year, from: date)
        let month = calendar.component(.month, from: date)
        return String(format: "%04d-%02d", year, month)
    }

    static func tierRank(_ tier: String?) -> Int {
        switch tier {
        case "bronze": return 1
        case "silver": return 2
        case "gold": return 3
        default: return 0
        }
    }

    static func normalizeMonthRerollEntry(_ entry: MonthRerollEntry?) -> MonthRerollEntry {
        entry ?? .empty
    }

    static func normalizeMonthlyChallengeRerolls(_ raw: [String: MonthRerollEntry]) -> [String: MonthRerollEntry] {
        raw.mapValues { normalizeMonthRerollEntry($0) }
    }

    static func evaluateMonthlyChallenges(
        sessions: [SwimSession],
        monthKey: String = getMonthKey(),
        rerolls: [String: MonthRerollEntry] = [:],
        intensity: Double = 1
    ) -> MonthlyChallengeState {
        let definitions = generateMonthlyChallenges(
            sessions: sessions,
            monthKey: monthKey,
            rerolls: rerolls,
            intensity: intensity
        )
        let monthSessions = sessions.filter { $0.date.hasPrefix(monthKey) }

        let challenges = definitions.map { def -> MonthlyChallenge in
            let current = measureChallenge(type: def.type, monthSessions: monthSessions, monthKey: monthKey)
            return MonthlyChallenge(
                id: def.id,
                type: def.type,
                monthKey: def.monthKey,
                target: def.target,
                tierIndex: def.tierIndex,
                current: current,
                completed: current >= def.target
            )
        }

        let completedCount = challenges.filter(\.completed).count
        let tier: String?
        switch completedCount {
        case 3...: tier = "gold"
        case 2: tier = "silver"
        case 1: tier = "bronze"
        default: tier = nil
        }

        let earnedAt = tier != nil && !monthSessions.isEmpty ? monthSessions.last?.date : nil
        return MonthlyChallengeState(
            monthKey: monthKey,
            challenges: challenges,
            completedCount: completedCount,
            tier: tier,
            earnedAt: earnedAt
        )
    }

    static func getMonthlyTierUpgrade(
        sessionsBefore: [SwimSession],
        sessionsAfter: [SwimSession],
        monthKey: String = getMonthKey(),
        rerolls: [String: MonthRerollEntry] = [:],
        intensity: Double = 1
    ) -> (monthKey: String, tier: String, fromTier: String?, completedCount: Int, earnedAt: String?)? {
        let before = evaluateMonthlyChallenges(
            sessions: sessionsBefore,
            monthKey: monthKey,
            rerolls: rerolls,
            intensity: intensity
        )
        let after = evaluateMonthlyChallenges(
            sessions: sessionsAfter,
            monthKey: monthKey,
            rerolls: rerolls,
            intensity: intensity
        )
        if tierRank(after.tier) > tierRank(before.tier), let tier = after.tier {
            return (monthKey, tier, before.tier, after.completedCount, after.earnedAt)
        }
        return nil
    }

    static func getMonthlyChallengeHistory(
        sessions: [SwimSession],
        previewMonthlyMedals: Bool = false,
        monthlyChallengeRerolls: [String: MonthRerollEntry] = [:],
        intensity: Double = 1
    ) -> [MonthlyChallengeState] {
        let months = Array(Set(sessions.map { String($0.date.prefix(7)) })).sorted(by: >)
        let real = months.compactMap { monthKey -> MonthlyChallengeState? in
            let state = evaluateMonthlyChallenges(
                sessions: sessions,
                monthKey: monthKey,
                rerolls: monthlyChallengeRerolls,
                intensity: intensity
            )
            return state.tier == nil ? nil : state
        }

        guard previewMonthlyMedals else { return real }

        let preview = getPreviewMonthlyMedalHistory(sessions: sessions)
        let realKeys = Set(real.map(\.monthKey))
        return (preview.filter { !realKeys.contains($0.monthKey) } + real)
            .sorted { $0.monthKey > $1.monthKey }
    }

    static func getMonthRerollOverrides(_ monthKey: String, rerolls: [String: MonthRerollEntry]) -> [String: String] {
        normalizeMonthRerollEntry(rerolls[monthKey]).overrides
    }

    static func getFreeMonthlyRerollsUsed(_ monthKey: String, rerolls: [String: MonthRerollEntry]) -> Int {
        normalizeMonthRerollEntry(rerolls[monthKey]).freeUses
    }

    static func hasRerollAvailability(
        monthKey: String,
        rerolls: [String: MonthRerollEntry],
        credits: Int,
        freeLimit: Int = 1
    ) -> Bool {
        getFreeMonthlyRerollsUsed(monthKey, rerolls: rerolls) < freeLimit || credits > 0
    }

    static func createMonthlyChallengeReroll(
        sessions: [SwimSession],
        monthKey: String,
        tierIndex: Int,
        rerolls: [String: MonthRerollEntry] = [:]
    ) -> (tierIndex: Int, type: String)? {
        guard tierIndex >= 0, tierIndex <= 2 else { return nil }
        let base = generateMonthlyChallenges(
            sessions: sessions,
            monthKey: monthKey,
            rerolls: rerolls,
            intensity: 1
        )
        let lockedTypes = base.enumerated().filter { $0.offset != tierIndex }.map(\.element.type)
        let currentType = base[tierIndex].type
        let candidates = challengeTypes.filter { !lockedTypes.contains($0) && $0 != currentType }
        let salt = getMonthRerollOverrides(monthKey, rerolls: rerolls).count
        guard let type = pickRerollType(monthKey: monthKey, tierIndex: tierIndex, candidates: candidates, salt: salt) else {
            return nil
        }
        return (tierIndex, type)
    }

    static func canRerollMonthlyChallenge(
        sessions: [SwimSession],
        monthKey: String,
        tierIndex: Int,
        rerolls: [String: MonthRerollEntry] = [:],
        credits: Int = 0,
        intensity: Double = 1,
        freeLimit: Int = 1
    ) -> Bool {
        guard hasRerollAvailability(monthKey: monthKey, rerolls: rerolls, credits: credits, freeLimit: freeLimit) else {
            return false
        }
        let state = evaluateMonthlyChallenges(
            sessions: sessions,
            monthKey: monthKey,
            rerolls: rerolls,
            intensity: intensity
        )
        guard tierIndex >= 0, tierIndex < state.challenges.count else { return false }
        let challenge = state.challenges[tierIndex]
        guard !challenge.completed else { return false }
        return createMonthlyChallengeReroll(
            sessions: sessions,
            monthKey: monthKey,
            tierIndex: tierIndex,
            rerolls: rerolls
        ) != nil
    }

    static func applyMonthlyChallengeReroll(
        data: SwimData,
        monthKey: String,
        tierIndex: Int,
        mascotId: String
    ) -> SwimData? {
        let credits = data.challengeRerollCredits
        let gameplay = MascotConstants.gameplay(mascotId)
        guard canRerollMonthlyChallenge(
            sessions: data.sessions,
            monthKey: monthKey,
            tierIndex: tierIndex,
            rerolls: data.monthlyChallengeRerolls,
            credits: credits,
            intensity: gameplay.challengeIntensity,
            freeLimit: gameplay.freeMonthlyRerolls
        ) else {
            return nil
        }

        guard let override = createMonthlyChallengeReroll(
            sessions: data.sessions,
            monthKey: monthKey,
            tierIndex: tierIndex,
            rerolls: data.monthlyChallengeRerolls
        ) else {
            return nil
        }

        var next = data
        var monthEntry = normalizeMonthRerollEntry(next.monthlyChallengeRerolls[monthKey])
        let useFree = monthEntry.freeUses < gameplay.freeMonthlyRerolls
        if !useFree && credits < 1 { return nil }

        monthEntry.overrides[String(override.tierIndex)] = override.type
        if useFree {
            monthEntry.freeUses += 1
        } else {
            next.challengeRerollCredits -= 1
        }
        next.monthlyChallengeRerolls[monthKey] = monthEntry
        return next
    }

    private static func getPreviewMonthlyMedalHistory(sessions: [SwimSession], date: Date = Date()) -> [MonthlyChallengeState] {
        ["gold", "silver", "bronze"].enumerated().map { index, tier in
            let calendar = Calendar.current
            let monthDate = calendar.date(byAdding: .month, value: -index, to: date) ?? date
            return buildPreviewMonthlyMedal(sessions: sessions, monthKey: getMonthKey(monthDate), tier: tier)
        }
    }

    private static func buildPreviewMonthlyMedal(
        sessions: [SwimSession],
        monthKey: String,
        tier: String
    ) -> MonthlyChallengeState {
        let state = evaluateMonthlyChallenges(sessions: sessions, monthKey: monthKey)
        let completedCount = tier == "gold" ? 3 : tier == "silver" ? 2 : 1
        let challenges = state.challenges.enumerated().map { index, challenge in
            MonthlyChallenge(
                id: challenge.id,
                type: challenge.type,
                monthKey: challenge.monthKey,
                target: challenge.target,
                tierIndex: challenge.tierIndex,
                current: index < completedCount ? challenge.target : max(0, Int(Double(challenge.target) * 0.4)),
                completed: index < completedCount
            )
        }
        let earnedDay = String(format: "%02d", min(28, 10 + completedCount * 5))
        return MonthlyChallengeState(
            monthKey: monthKey,
            challenges: challenges,
            completedCount: completedCount,
            tier: tier,
            earnedAt: "\(monthKey)-\(earnedDay)",
            isPreview: true
        )
    }

    private static func pickRerollType(
        monthKey: String,
        tierIndex: Int,
        candidates: [String],
        salt: Int
    ) -> String? {
        guard !candidates.isEmpty else { return nil }
        var seed = hashMonth("\(monthKey):reroll:\(tierIndex):\(salt)")
        if seed == 0 { seed = 1 }
        seed = (seed &* 1_103_515_245 &+ 12_345) & 0x7fff_ffff
        return candidates[seed % candidates.count]
    }

    static func getMonthlyShortfallPenalty(
        sessions: [SwimSession],
        uploadMonthKey: String,
        mascotId: String?,
        rerolls: [String: MonthRerollEntry],
        settledMonths: [String: MonthlySettlement]
    ) -> MonthlyShortfallPenalty? {
        guard let mascotId, !uploadMonthKey.isEmpty else { return nil }
        let gameplay = MascotConstants.gameplay(mascotId)
        guard let requiredTier = gameplay.requiredMonthlyTier, gameplay.monthlyPenaltyCoins > 0 else { return nil }

        let parts = uploadMonthKey.split(separator: "-").compactMap { Int($0) }
        guard parts.count == 2 else { return nil }
        var components = DateComponents(year: parts[0], month: parts[1], day: 1)
        components.month = (components.month ?? 1) - 1
        guard let prevDate = Calendar.current.date(from: components) else { return nil }
        let prevMonthKey = getMonthKey(prevDate)

        if settledMonths[prevMonthKey] != nil { return nil }
        if !sessions.contains(where: { $0.date.hasPrefix(prevMonthKey) }) { return nil }

        let state = evaluateMonthlyChallenges(
            sessions: sessions,
            monthKey: prevMonthKey,
            rerolls: rerolls,
            intensity: gameplay.challengeIntensity
        )
        if tierRank(state.tier) >= tierRank(requiredTier) { return nil }

        return MonthlyShortfallPenalty(
            monthKey: prevMonthKey,
            coins: gameplay.monthlyPenaltyCoins,
            achievedTier: state.tier,
            requiredTier: requiredTier,
            mascotId: mascotId
        )
    }

    static func generateMonthlyChallenges(
        sessions: [SwimSession],
        monthKey: String,
        rerolls: [String: MonthRerollEntry] = [:],
        intensity: Double = 1
    ) -> [ChallengeDefinition] {
        generateMonthlyChallengeDefinitions(
            sessions: sessions,
            monthKey: monthKey,
            rerolls: rerolls,
            intensity: intensity
        )
    }

    static func hasMonthlyChallengeReroll(_ monthKey: String, rerolls: [String: MonthRerollEntry]) -> Bool {
        !getMonthRerollOverrides(monthKey, rerolls: rerolls).isEmpty
    }

    static func getMonthlyMedalsForYear(
        sessions: [SwimSession],
        year: Int = Calendar.current.component(.year, from: Date()),
        monthlyChallengeRerolls: [String: MonthRerollEntry] = [:],
        intensity: Double = 1
    ) -> [YearMonthMedal] {
        (1...12).map { month in
            let monthKey = String(format: "%04d-%02d", year, month)
            let state = evaluateMonthlyChallenges(
                sessions: sessions,
                monthKey: monthKey,
                rerolls: monthlyChallengeRerolls,
                intensity: intensity
            )
            return YearMonthMedal(
                monthKey: monthKey,
                month: month,
                tier: state.tier,
                completedCount: state.completedCount,
                challenges: state.challenges,
                earnedAt: state.earnedAt,
                hasSessions: sessions.contains { $0.date.hasPrefix(monthKey) }
            )
        }
    }

    static func getPreviewMonthlyMedalHistoryForTesting(
        sessions: [SwimSession] = [],
        date: Date = Date()
    ) -> [MonthlyChallengeState] {
        getPreviewMonthlyMedalHistory(sessions: sessions, date: date)
    }

    struct ChallengeDefinition {
        var id: String
        var type: String
        var monthKey: String
        var target: Int
        var tierIndex: Int
    }

    private static func generateMonthlyChallengeDefinitions(
        sessions: [SwimSession],
        monthKey: String,
        rerolls: [String: MonthRerollEntry],
        intensity: Double
    ) -> [ChallengeDefinition] {
        let stats = computeRecentMonthlyStats(sessions: sessions, beforeMonthKey: monthKey)
        let types = pickChallengeTypes(monthKey: monthKey)
        let challenges = types.enumerated().map { index, type in
            ChallengeDefinition(
                id: "\(monthKey)_\(type)",
                type: type,
                monthKey: monthKey,
                target: buildTarget(type: type, stats: stats, tierIndex: index, intensity: intensity),
                tierIndex: index
            )
        }
        return applyRerollOverrides(challenges: challenges, monthEntry: rerolls[monthKey], stats: stats, intensity: intensity)
    }

    private static func computeRecentMonthlyStats(sessions: [SwimSession], beforeMonthKey: String) -> (
        avgSessions: Int, avgDistance: Int, avgKcal: Int, avgStreak: Int, avgWeeks: Int
    ) {
        var byMonth: [String: (sessions: Int, distanceM: Int, activeKcal: Int)] = [:]
        for session in sessions {
            let monthKey = String(session.date.prefix(7))
            guard monthKey < beforeMonthKey else { continue }
            var bucket = byMonth[monthKey] ?? (0, 0, 0)
            bucket.sessions += 1
            bucket.distanceM += session.metrics.distanceM ?? 0
            bucket.activeKcal += session.metrics.activeKcal ?? 0
            byMonth[monthKey] = bucket
        }

        let months = byMonth.keys.sorted().suffix(3)
        guard !months.isEmpty else {
            return (4, 6000, 2500, 2, 2)
        }

        var totals = (sessions: 0, distanceM: 0, activeKcal: 0)
        for month in months {
            let bucket = byMonth[month]!
            totals.sessions += bucket.sessions
            totals.distanceM += bucket.distanceM
            totals.activeKcal += bucket.activeKcal
        }
        let count = months.count
        return (
            max(1, Int(round(Double(totals.sessions) / Double(count)))),
            max(1000, Int(round(Double(totals.distanceM) / Double(count)))),
            max(500, Int(round(Double(totals.activeKcal) / Double(count)))),
            2,
            2
        )
    }

    private static func buildTarget(type: String, stats: (avgSessions: Int, avgDistance: Int, avgKcal: Int, avgStreak: Int, avgWeeks: Int), tierIndex: Int, intensity: Double) -> Int {
        let stretch = (1 + Double(tierIndex) * 0.12) * intensity
        switch type {
        case "sessions":
            return max(2, Int(ceil(Double(stats.avgSessions) * stretch)))
        case "distance":
            return max(Int(ceil(4000 * intensity / 500) * 500), Int(ceil(Double(stats.avgDistance) * stretch / 500) * 500))
        case "kcal":
            return max(Int(ceil(1500 * intensity / 250) * 250), Int(ceil(Double(stats.avgKcal) * stretch / 250) * 250))
        case "streak":
            return max(2, min(7, stats.avgStreak + tierIndex))
        case "active_weeks":
            return max(2, min(4, stats.avgWeeks + tierIndex - 1))
        default:
            return 1
        }
    }

    private static func pickChallengeTypes(monthKey: String) -> [String] {
        var pool = challengeTypes
        var picked: [String] = []
        var seed = hashMonth(monthKey)
        for _ in 0..<3 {
            seed = (seed &* 1_103_515_245 &+ 12_345) & 0x7fff_ffff
            let index = seed % pool.count
            picked.append(pool.remove(at: index))
        }
        return picked
    }

    private static func hashMonth(_ monthKey: String) -> Int {
        monthKey.unicodeScalars.reduce(0) { ($0 &* 31 &+ Int($1.value)) | 0 }
    }

    private static func applyRerollOverrides(
        challenges: [ChallengeDefinition],
        monthEntry: MonthRerollEntry?,
        stats: (avgSessions: Int, avgDistance: Int, avgKcal: Int, avgStreak: Int, avgWeeks: Int),
        intensity: Double
    ) -> [ChallengeDefinition] {
        let overrides = normalizeMonthRerollEntry(monthEntry).overrides
        return challenges.enumerated().map { index, challenge in
            guard let type = overrides[String(index)] else { return challenge }
            return ChallengeDefinition(
                id: "\(challenge.monthKey)_\(type)",
                type: type,
                monthKey: challenge.monthKey,
                target: buildTarget(type: type, stats: stats, tierIndex: index, intensity: intensity),
                tierIndex: index
            )
        }
    }

    private static func measureChallenge(type: String, monthSessions: [SwimSession], monthKey: String) -> Int {
        switch type {
        case "sessions":
            return monthSessions.count
        case "distance":
            return monthSessions.compactMap(\.metrics.distanceM).reduce(0, +)
        case "kcal":
            return monthSessions.compactMap(\.metrics.activeKcal).reduce(0, +)
        case "streak":
            return maxConsecutiveDaysInMonth(sessions: monthSessions, monthKey: monthKey)
        case "active_weeks":
            return activeWeeksInMonth(monthSessions: monthSessions, monthKey: monthKey)
        default:
            return 0
        }
    }

    private static func maxConsecutiveDaysInMonth(sessions: [SwimSession], monthKey: String) -> Int {
        let dates = Array(Set(sessions.filter { $0.date.hasPrefix(monthKey) }.map(\.date))).sorted()
        guard !dates.isEmpty else { return 0 }
        var maxStreak = 1
        var streak = 1
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        for index in 1..<dates.count {
            guard let prev = formatter.date(from: dates[index - 1]),
                  let curr = formatter.date(from: dates[index]) else { continue }
            let diff = Calendar.current.dateComponents([.day], from: prev, to: curr).day ?? 0
            streak = diff == 1 ? streak + 1 : 1
            maxStreak = max(maxStreak, streak)
        }
        return maxStreak
    }

    private static func activeWeeksInMonth(monthSessions: [SwimSession], monthKey: String) -> Int {
        Set(monthSessions.filter { $0.date.hasPrefix(monthKey) }.map { weekKey(for: $0.date) }).count
    }

    private static func weekKey(for dateStr: String) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        guard let date = formatter.date(from: dateStr) else { return dateStr }
        let calendar = Calendar.current
        let weekday = calendar.component(.weekday, from: date)
        let diff = calendar.component(.day, from: date) - weekday + (weekday == 1 ? -6 : 1)
        guard let monday = calendar.date(byAdding: .day, value: diff - calendar.component(.day, from: date), to: date) else {
            return dateStr
        }
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: monday)
    }
}
